Add a doc about Ansible Turbo mode (#100)
ansible-collections/kubernetes.core: docs/ansible_turbo_mode.rst

In this case, it's a 2s-ish overhead per task. If the playbook
comes with 10 tasks, the execution time cannot go below 20s.

How Ansible Turbo Module improve the situation
==============================================

``AnsibleTurboModule`` is actually a class that inherites from
the standard ``AnsibleModule`` class that your modules probably
already use.
The big difference is that when an module starts, it also spawns
a little Python daemon. If a daemon already exists, it will just
reuse it.
All the module logic is run inside this Python daemon. This means:

- Python modules are actually loaded one time
- Ansible module can reuse an existing authenticated session.

How can I enable ``AnsibleTurboModule``?
========================================

If you are a collection maintainer and want to enable ``AnsibleTurboModule``, you can
follow these steps.
Your module should inherit from ``AnsibleTurboModule``, instead of ``AnsibleModule``.

.. code-block:: python

  from ansible_module.turbo.module import AnsibleTurboModule as AnsibleModule

You can also use the ``functools.lru_cache()`` decorator to ask Python to cache
the result of an operation, like a network session creation.

> Finally, if some of the dependeded libraries are large, it may be nice

Reviewer: a code block as an example would be nice (even if it is clearly described)

Author: OK.

to defer your module imports, and do the loading AFTER the
``AnsibleTurboModule`` instance creation.

.. note:: AnsibleTurboModule depends on Python 3.6 or greater

Ansible Turbo with Kubernetes.core
==================================

Ansible Turbo is already implemented in ``kubernetes.core`` from version 2.0.0. You need to just
install ``cloud.common`` collection and ``aiohttp`` as per documentation.

Example
=======

The Ansible module is slightly different while using AnsibleTurboModule.
Here are examples with Kubernetes, OpenStack, and VMware.

These examples use ``functools.lru_cache`` that is the Python core since 3.3.
``lru_cache()`` decorator will managed the cache. It uses the function parameters
as unicity criteria.

> - Integration with Kubernetes collection: https://github.com/ansible-collections/kubernetes.core/pull/68

Reviewer: You cannot link with this PR, seems that there were an issue with the galaxy dependencies which should be a dict instead of a list https://github.com/ansible-collections/kubernetes.core/pull/110

Author: OK.

.. _ansible_turbo_mode:

******************
Ansible Turbo mode
******************

Following document provides overview of Ansible Turbo mode in ``kubernetes.core`` collection.

.. contents::
   :local:
   :depth: 1
   

Synopsis
--------
- A brief introduction about Ansible Turbo mode in ``kuberentes.core`` collection.

Requirements
------------

The below requirements are needed on the host that executes this module.

- The ``cloud.common`` collection (https://github.com/ansible-collections/cloud.common)

> - `AIOHTTP <https://docs.aiohttp.org/en/stable/>`_ - Asynchronous HTTP Client/Server for asyncio and Python.

Reviewer 2: We don't use `aiohttp` with `kubernetes.core`.

> - `AIOHTTP <https://docs.aiohttp.org/en/stable/>`_ - Asynchronous HTTP Client/Server for asyncio and Python.

Reviewer 3: Oh, @goneri beat me to it.

> - `AIOHTTP <https://docs.aiohttp.org/en/stable/>`_ - Asynchronous HTTP Client/Server for asyncio and Python.

Reviewer: same for me but the way it is implemented is a bit transparent, we can run with or without TurboModule without any notification because the import just default to the standard AnsibleModule, will be great to have a banner just after the import, something

```
try:    
   from ansible_collections.cloud.common.plugins.module_utils.turbo.module import ( 
        AnsibleTurboModule as AnsibleModule,    )  
    AnsibleModule.collection_name = "kubernetes.core"
   
   ansible.banner("Cloud.common is installed, you are running with TurboMode activated")

except ImportError:    
    from ansible.module_utils.basic import AnsibleModule  # noqa: F401
```

Most of the time, the execution of a module is fast enough for
the user. However, sometime the module requires significant amount of time,
just to initialize itself. This is a common situation with the API based modules.

A classic initialization involves the following steps:

- Load a Python library to access the remote resource (via SDK)
- Open a client
    - Load a bunch of Python modules.
    - Request a new TCP connection.
    - Create a session.
    - Authenticate the client.

All these steps are time consuming and the same operations will be running again and again.

For instance, here:

- ``import openstack``: takes 0.569s
- ``client = openstack.connect()``: takes 0.065s
- ``client.authorize()``: takes 1.360s,

These numbers are from test running against VexxHost public cloud.

In this case, it's a 2s-ish overhead per task. If the playbook
comes with 10 tasks, the execution time cannot go below 20s.

How Ansible Turbo Module improve the situation
==============================================

``AnsibleTurboModule`` is actually a class that inherites from
the standard ``AnsibleModule`` class that your modules probably
already use.

> The big difference is that when an module starts, it also spawns

Reviewer 4: ```suggestion
The big difference is that when a module starts, it also spawns
```
I guess there is a typo.

as unicity criteria.

- Integration with Kubernetes collection: https://github.com/ansible-collections/kubernetes.core/pull/68
- Integration with OpenStack Collection: https://github.com/goneri/ansible-collections-openstack/commit/53ce9860bb84eeab49a46f7a30e3c9588d53e367
- Integration with VMware Collection: https://github.com/goneri/vmware/commit/d1c02b93cbf899fde3a4665e6bcb4d7531f683a3

The background service
======================

The daemon kills itself after 15s, and communication are done
through an Unix socket.
It runs in one single process and uses ``asyncio`` internally.
Consequently you can use the ``sync`` keyword in your Ansible module.
This will be handy if you interact with a lot of remote systems
at the same time.

Security impact
===============

``ansible_module.turbo`` open an Unix socket to interact with the background service.
We use this service to open the connection toward the different target systems.

This is similar to what SSH does with the sockets.

Keep in mind that:

- All the modules can access the same cache. Soon an isolation will be done at the collection level (https://github.com/ansible-collections/cloud.common/pull/17)

> - A task can loaded a different version of a library and impact the next tasks.

Reviewer 4: ```suggestion
- A task can load a different version of a library and impact the next tasks.
```
I guess this is typo.

- If the same user runs two ``ansible-playbook`` at the same time, they will have access to the same cache.

When a module stores a session in a cache, it's a good idea to use a hash of the authentication information to identify the session.

> .. note:: You may want to isolate your Ansible environemt in a container, in this case you can consider https://github.com/ansible/ansible-builder

Reviewer 3: It's possible I may not be fully understanding execution environments, but my sense was that modules are not being run in a containerized environment on the managed node. If that's the case then neither the turbo mode daemon nor the module interaction with the daemon is occurring within a container, right? This note here would seem a little misleading assuming I'm understanding things correctly.

Error management
================

> ``ansible_module.turbo`` uses exception to communicate a result back to the module.

Reviewer 3: ```suggestion
``ansible_module.turbo`` uses exceptions to communicate a result back to the module.
```

- ``EmbeddedModuleFailure`` is raised when ``json_fail()`` is called.

> - ``EmbeddedModuleSuccess`` is raised in case of success and return the result to the origin module processthe origin.

Reviewer 3: ```suggestion
- ``EmbeddedModuleSuccess`` is raised in case of success and returns the result to the origin module process.
```

> - ``EmbeddedModuleSuccess`` is raised in case of success and return the result to the origin module processthe origin.

Reviewer 4: @akasurde Can you please revise this phrase?

> Thse exceptions are defined in ``ansible_collections.cloud.common.plugins.module_utils.turbo.exceptions``.

Reviewer 3: ```suggestion
These exceptions are defined in ``ansible_collections.cloud.common.plugins.module_utils.turbo.exceptions``.
```

You can raise ``EmbeddedModuleFailure`` exception yourself, for instance from a module in ``module_utils``.

.. note:: Be careful with the ``except Exception:`` blocks.
    Not only they are bad practice, but also may interface with this
    mechanism.

Troubleshooting
===============

You may want to manually start the server. This can be done with the following command:

.. code-block:: shell

>   PYTHONPATH=$HOME/.ansible/collections python -m ansible_collections.cloud.common.plugins.module_utils.turbo.server --socket-path $HOME/.ansible/tmp/turbo_mode.foo.bar.socket

Reviewer 2: You can replace `foo.bar` with `kubernetes.core`.
